src/FAudio_gstreamer.c

```diff
@@ -0,0 +1,850 @@
+	}
+
+	/* turn off sync so we can pull data without waiting for it to "play" in realtime */
+	g_value_init(&gval, G_TYPE_BOOLEAN);
+	g_value_set_boolean(&gval, FALSE);
+	g_object_set_property(G_OBJECT(result->dst), "sync", &gval);
+	g_value_set_boolean(&gval, TRUE);
+	g_object_set_property(G_OBJECT(result->dst), "async", &gval);
+
+	/* Compile the pipeline, finally. */
+	if (!gst_pad_set_active(result->srcpad, TRUE))
+	{
+		LOG_ERROR(
+			pSourceVoice->audio,
+			"%s",
+			"Unable to activate srcpad"
+		)
+		goto free_without_bin;
+	}
+
+	gst_bin_add_many(
+		GST_BIN(result->pipeline),
+		decoder,
+		result->resampler,
+		converter,
+		result->dst,
+		NULL
+	);
+
+	decoder_sink = gst_element_get_static_pad(decoder, "sink");
+	if (!decoder_sink)
+	{
+		LOG_ERROR(
+			pSourceVoice->audio,
+			"%s",
+			"Unable to get decoder sink pad"
+		)
+		goto free_with_bin;
+	}
+
+	if (gst_pad_link(result->srcpad, decoder_sink) != GST_PAD_LINK_OK)
+	{
+		LOG_ERROR(
+			pSourceVoice->audio,
+			"%s",
+			"Unable to get link our src pad to decoder sink pad"
+		)
+		gst_object_unref(decoder_sink);
+		goto free_with_bin;
+	}
+
+	gst_object_unref(decoder_sink);
+
+	if (!gst_element_link_many(
+		result->resampler,
+		converter,
+		result->dst,
+		NULL))
+	{
+		LOG_ERROR(
+			pSourceVoice->audio,
+			"%s",
+			"Unable to get link pipeline"
+		)
+		goto free_with_bin;
+	}
+
+	/* send stream-start */
+	event = gst_event_new_stream_start("faudio/gstreamer");
+	if (!event)
+	{
+		LOG_ERROR(
+			pSourceVoice->audio,
+			"Unable to create stream_start event; is %u-bit gstreamer installed?",
+			sizeof(void *) * 8
+		)
+		goto free_with_bin;
+	}
+
+	if (!gst_pad_push_event(result->srcpad, event))
+	{
+		LOG_ERROR(
+			pSourceVoice->audio,
+			"%s",
+			"Pushing stream_start event failed"
+		)
+		goto free_with_bin;
+	}
+
+	/* Prepare the input format */
+	if (type == FAUDIO_FORMAT_WMAUDIO3)
+	{
+		const FAudioWaveFormatExtensible *wfx =
+			(FAudioWaveFormatExtensible*) pSourceVoice->src.format;
+		extradata = (uint8_t*) &wfx->Samples;
+		codec_data_size = pSourceVoice->src.format->cbSize;
+	}
+	else if (type == FAUDIO_FORMAT_WMAUDIO2)
+	{
+		FAudio_zero(fakeextradata, sizeof(fakeextradata));
+		fakeextradata[4] = 31;
+
+		extradata = fakeextradata;
+		codec_data_size = sizeof(fakeextradata);
+	}
+	else
+	{
+		FAudio_assert(0 && "Unrecognized WMA format!");
+	}
+	codec_data = gst_buffer_new_and_alloc(codec_data_size);
+	gst_buffer_fill(codec_data, 0, extradata, codec_data_size);
+	caps = gst_caps_new_simple(
+		"audio/x-wma",
+		"wmaversion",	G_TYPE_INT, version,
+		"bitrate",	G_TYPE_INT, pSourceVoice->src.format->nAvgBytesPerSec * 8,
+		"channels",	G_TYPE_INT, pSourceVoice->src.format->nChannels,
+		"rate",		G_TYPE_INT, pSourceVoice->src.format->nSamplesPerSec,
+		"block_align",	G_TYPE_INT, pSourceVoice->src.format->nBlockAlign,
+		"depth",	G_TYPE_INT, pSourceVoice->src.format->wBitsPerSample,
+		"codec_data",	GST_TYPE_BUFFER, codec_data,
+		NULL
+	);
```

Comment:
Instead of unreffing the buffer in all the different return paths below, just unref it directly here after creating the caps